src/dexpo_socket.cpp

```diff
@@ -0,0 +1,142 @@
+#include "dexpo_socket.hpp"
+#include <algorithm>
+#include <cstdio>
+#include <cstring>
+#include <iostream>
+#include <ostream>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+#include <unistd.h>
+
+template <typename t>
+void
+handle_error (t var)
+{
+  if (var == -1)
+    {
+      throw;
+    }
+}
+
+/**
+ * Connect to socket.
+ *
+ * If socket does not exist:
+ * - Create a socket file descriptor
+ * - Bind socket to name
+ * - Listen (specify max connection number)
+ * - Accept
+ *
+ * If socket exists:
+ * - Create a socket file descriptor
+ * - Connect to socket
+ */
+dexpo_socket::dexpo_socket ()
+{
+  struct sockaddr_un sock_name = {};
+  sock_name.sun_family = AF_UNIX;
+  // Safely putting socket name into a buffer
+  std::strncpy (sock_name.sun_path, SOCKET_PATH,
+                sizeof (sock_name.sun_path) - 1);
+
+  // XXX Make socket abstract
+  // sock_name.sun_path[0] = '\0';
+
+  // Create a socket file descriptor
+  this->fd = socket (AF_UNIX, SOCK_STREAM, 0);
+  handle_error (this->fd);
+
+  // Trick compiler into thinking that we pass pointer to sockaddr struct
+  // https://stackoverflow.com/questions/21099041
+  auto *sock_addr = reinterpret_cast<struct sockaddr *> (&sock_name);
+
+  // Connect to named socket
+  int e = connect (this->fd, sock_addr, sizeof (sock_name));
+  if (e == -1) // Socket doesn't exist
```

Comment:
Why not add it to handle_error?

Reply:
handle_error will `throw`, but here we need to handle the error differently
I need to add custom error types and improve error handling in general